docs/CodingStandards.md

```diff
@@ -0,0 +1,162 @@
+it is a single line, including comments. Otherwise the body should be braced.
+
+### Redundant comparisons
+
+Comparison of a bool with `false` or a pointer with `nullptr` is redundant and should
+be avoided. Use `!` instead where appropriate.
+
+However a comparison of an integer with 0 is useful. Consider keeping that.
+
+### Redundant parentheses
+
+Certain cases of theoretically redundant parentheses around a subexpression are needed
+because compilers warn about them otherwise, for the good reason that C/C++ operator
+precedence makes it difficult to understand and easy to get wrong (in particular
+`&` `|` `^` vs comparison operators, and `&&` vs `||`). However redundant parentheses
+are not needed in cases that compilers do not warn about, in particular a comparison
+inside `&&` or `||`, or a comparison used as the condition of `?:` or in an assignment
+or `return`. Avoid redundant parentheses in these cases, unless really needed for
+clarity.
+
+### Comments
+
+Despite the look of much LLVM code, the LLVM coding standard does actually recommend
+placing a comment at the top of a function/method. :-)
+
+In LLPC, that is mandatory. Parameters should be commented in the main comment at
+the top of the function like this:
+
+```
+// @param hatstand : [in/out] The Hatstand object to modify
+```
+
+A tag of `[in]` is unnecessary, and is assumed for a reference or pointer parameter
+where no tag is given. But for a pointer or reference that is not just an input,
+specify a tag of `[out]` or `[in/out]` as appropriate.
+
+Parameter comments should not be placed on the parameter line itself and aligned
+(as happens in PAL style), to avoid spuriously large diffs when a change or addition
+of one parameter means all the others need realigning.
+
+Similarly, a bunch of fields or variables should not have aligned inline comments.
+Instead, put each field's comment on the line above the field.
+
+### clang-format
+
+The LLPC coding standard uses a clang-format configuration based on the LLVM one,
+with the following changes:
+
+- 120 column limit (rather than 80)
+- A short function is allowed in a single line only if it is inline (inside a class)
+- A short lambda is allowed in a single line only if it is empty
+
+The #includes at the top of the file must be ordered as:
+
+- LLPC files, in alphabetical order
+- SPIRV and pal files, in alphabetical oder
+- llvm files, in alphabetical order
+- system header files used with angle brackets, in alphabetical order.
+
+Clang-format enforces this order, if you run clang-format on that part of the file.
+
+You should use clang-format to format code modified or added in your change, but do not
+reformat any other code. Reviewers should reject a change that has spurious reformatting
+outside the lines affected by the substantive change, unless the whole point of the change
+is the reformatting. You can use
+[git clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format)
```

Comment:
Maybe git-clang-format has changed since I last downloaded it, but I thought it formatted the entire file for every file that is changed, not just the lines that were change in those files.  So git clang format may not do what you want.

Reply:
If we are changing a file that has not been formatted to match the new standard, should be do a PR that just formats that file first, and then make our change?

Will there a mass reformatting taking place?

Reply:
Yes there will be a mass reformatting (together with a hacked clang-tidy and a script that you can use in git filter-branch on your own local changes so you can rebase, or you can ask me to do that to your local changes).

Reply:
Perfect.  I'm in favour of a mass rebase.  Short-term pain for long-term gain.

Reply:
git-clang-format can reformat just the lines that you've changed: https://github.com/llvm/llvm-project/blob/master/clang/tools/clang-format/git-clang-format#L38

I usually run `git clang-format HEAD^` after committing.